File: tools/verify_docs.mjs
// /docs(README.md から生成)の検証。
//   1) Markdown変換の単体チェック(表・コード・エスケープ・インライン記法)
//   2) 実際の /docs が 200 で、READMEの主要セクションがHTMLになっているか
//   3) 目次リンクと見出しIDが全部一致しているか(リンク切れ検出)
//   4) README中のHTMLタグがエスケープされているか(コードブロック内の <div> 等)
//
// 使い方: node tools/verify_docs.mjs [baseUrl]
import { readFileSync } from "node:fs";
import { renderMarkdown, slugify } from "../lib/md.mjs";

const BASE = process.argv[2] || "http://localhost:3107";
let pass = 0;
let fail = 0;
const check = (label, ok, extra = "") => {
  if (ok) pass++;
  else fail++;
  console.log((ok ? "  ✅ " : "  ❌ ") + label + (extra ? "  " + extra : ""));
};

// ---------- 1) Markdown変換の単体チェック ----------
console.log("=== A) Markdown変換 ===");
const fixture = [
  "# 見出し1",
  "",
  "## 見出し2",
  "",
  "**太字** と *斜体* と `コード` と [リンク](https://example.com/a?b=1&c=2)。",
  "",
  "```bash",
  'echo "<div id=x>&amp;</div>"',
  "```",
  "",
  "| 列A | 列B |",
  "|---|---|",
  "| `x` | **y** |",
  "",
  "- 項目1",
  "- 項目2",
  "",
  "1. 手順1",
  "2. 手順2",
  "",
  "> 引用文",
  "",
  "---",
].join("\n");
const r = renderMarkdown(fixture);
const h = r.html;

check("見出しがh1/h2になる", h.includes("<h1 id=") && h.includes("<h2 id="));
check("太字/斜体/コード/リンクが変換される",
  h.includes("<strong>太字</strong>") && h.includes("<em>斜体</em>") &&
  h.includes("<code>コード</code>") && h.includes('<a href="https://example.com/a?b=1&amp;c=2"'));
check("コードブロックがpre/codeになり、中のHTMLタグがエスケープされる",
  h.includes("<pre class=\"code\" data-lang=\"bash\">") && h.includes("&lt;div id=x&gt;") && !h.includes("<div id=x>"));
check("表がtableになり、セル内のインライン記法も効く",
  h.includes("<table>") && h.includes("<th>列A</th>") && h.includes("<td><code>x</code></td>") && h.includes("<td><strong>y</strong></td>"));
check("箇条書きがul/olになる", h.includes("<ul><li>項目1</li>") && h.includes("<ol><li>手順1</li>"));
check("引用と区切りが変換される", h.includes("<blockquote>引用文</blockquote>") && h.includes("<hr>"));
check("生のMarkdownが残っていない(``` も |---| も無い)",
  !h.includes("```") && !/\|\s*---/.test(h));
check("目次用の見出しを拾っている(h1は除外してh2のみ)", r.toc.length === 1 && r.toc[0].level === 2,
  `toc=${JSON.stringify(r.toc.map((t) => t.level + ":" + t.text))}`);

// 見出しIDが重複しても壊れないこと
const dup = renderMarkdown("## 同じ\n\n## 同じ\n").html;
check("同じ見出しでもIDが重複しない", dup.includes('id="同じ"') && dup.includes('id="同じ-1"'));

// 記号が落ちて日本語が残ること
check("slugifyが日本語を残し記号を落とす", slugify("テスト(注) 「A/B」") === "テスト注-ab");

// 生HTMLを書いても素通ししない(エスケープされる)
const raw = renderMarkdown('<script>alert(1)</script>\n').html;
check("地の文の生タグもエスケープされる", raw.includes("&lt;script&gt;") && !raw.includes("<script>"));

// ---------- 2) 実際の /docs ----------
console.log("\n=== B) /docs の配信内容 ===");
const res = await fetch(BASE + "/docs", { headers: { "user-agent": "verify-docs/1.0" } });
const page = await res.text();
check("/docs が 200 で HTML を返す", res.status === 200 && /text\/html/.test(res.headers.get("content-type") || ""),
  `HTTP${res.status} ${(res.headers.get("content-type") || "").split(";")[0]}`);

const readme = readFileSync("C:/Users/maeba/Desktop/hikamani-captcha/README.md", "utf8");

// READMEの主要セクションがHTMLに載っているか(READMEの ## 見出しを全部確認)
const readmeH2 = [...readme.matchAll(/^##\s+(.+)$/gm)].map((m) => m[1].trim());
const missing = readmeH2.filter((t) => {
  const id = slugify(t.replace(/`/g, ""));
  return !page.includes(`id="${id}"`);
});
check(`READMEの全セクション(${readmeH2.length}個)がHTMLに入っている`, missing.length === 0, missing.slice(0, 3).join(" / "));

// 主要な中身(API・埋め込み・設定)が実際に載っているか
for (const key of ["/api/challenge", "/api/verify", "/api/consume", "/api/img/", "IP_QUOTA", "MIN_SOLVE_MS", "onSolved"]) {
  check(`本文に ${key} が出る`, page.includes(key));
}

// ---------- 3) 目次と見出しIDの整合 ----------
console.log("\n=== C) 目次リンクの整合 ===");
// ページ内の全アンカー(サイドバー目次・モバイルの折りたたみ目次)をまとめて検証する
const tocHrefs = [...page.matchAll(/href="#([^"]+)"/g)].map((m) => m[1]);
// アンカーの飛び先は見出しに限らない(#top はレイアウト要素)ので、ページ内の全idを見る
const ids = new Set([...page.matchAll(/\sid="([^"]+)"/g)].map((m) => m[1]));
const broken = tocHrefs.filter((t) => !ids.has(t));
check("目次リンクが全て存在する見出しを指す", tocHrefs.length > 5 && broken.length === 0,
  `${tocHrefs.length}本 / 切れ ${broken.length}${broken.length ? ": " + broken.slice(0, 3).join(",") : ""}`);
check("モバイル用の折りたたみ目次がある", page.includes('<details class="toc-m">') && page.includes("目次("));
check("固定ヘッダー分のスクロール余白がある", page.includes("scroll-margin-top"));

// ---------- 4) 描画の安全確認 ----------
console.log("\n=== D) 生成HTMLの安全性 ===");
// README内のコードブロックにあるタグが素のHTMLとして出ていないこと
check("README中の <div id=\"captcha\"> がエスケープされている", page.includes("&lt;div id=&quot;captcha&quot;&gt;"));
// 生成ページ自身のscriptは無し(静的な文書)
check("生成ページにscriptタグが無い", !/<script/i.test(page));
check("生成ページにstyleタグが1つだけある", (page.match(/<style/g) || []).length === 1);
check("HTMLが閉じている", page.trim().endsWith("</html>"));

console.log(`\n=== 合計: ${pass} PASS / ${fail} FAIL ===`);
process.exit(fail ? 1 : 0);

File: tools/verify-docs.py
""" /docs の描画を実測する(README.mdと一致しているか・崩れていないか)

README.md を唯一の情報源にして毎回レンダリングする設計なので、
「READMEの内容が全部出ているか」と「Markdownが崩れていないか」を見る。
"""
import re
import sys
import urllib.request

from playwright.sync_api import sync_playwright

BASE = "http://localhost:3107"
DOCS = BASE + "/docs"
SHOT = "C:/Users/maeba/Desktop/hikamani-captcha/.docs-shots"

OUT = []


def log(label, ok, extra=""):
    OUT.append((label, ok, extra))
    print(("OK   " if ok else "FAIL "), label, extra, flush=True)


def get(url):
    req = urllib.request.Request(url, headers={"user-agent": "verify-docs/1.0"})
    with urllib.request.urlopen(req, timeout=20) as r:
        return r.status, r.read().decode("utf-8", "replace")


# ---------- 1) 配信されているか ----------
st, html = get(DOCS)
log("/docs が200で返る", st == 200, f"HTTP{st} {len(html)}bytes")

readme = open("C:/Users/maeba/Desktop/hikamani-captcha/README.md", encoding="utf-8").read()

# ---------- 2) READMEの主要な見出しが出ているか ----------
h2 = re.findall(r"^## (.+)$", readme, re.M)
missing = [h for h in h2 if h.strip() not in html]
log("READMEの見出し(h2)が全部出ている", not missing, f"欠け={missing[:4]}" if missing else f"{len(h2)}個すべて")

# ---------- 3) 最新の追記が反映されているか(直近の変更を狙い撃ち) ----------
for probe, label in [
    ("0で無制限", "レート制限が既定オフの記述"),
    ("IP_QUOTA", "env表の IP_QUOTA"),
    ("HARD_MIN_SOLVE_MS", "回答が速すぎますの修正"),
    ("サイトキー", "複数サイトで足りないもの"),
    ("18 PASS / 0 FAIL", "埋め込み検証の結果"),
    ("20 PASS / 0 FAIL", "多層認証テストの結果"),
]:
    log(f"最新の記述: {label}", probe in html, f"'{probe}'" + ("" if probe in html else " が見つからない"))

# ---------- 4) Markdownの崩れ(表・コード・リンク) ----------
log("表がテーブルに変換されている", html.count("<table") >= 5, f"<table> {html.count('<table')}個")
log("コードがコードブロックに変換されている", html.count("<pre") >= 5, f"<pre> {html.count('<pre')}個")
log("生のMarkdown記号が残っていない(|で始まる行や```)", not re.search(r"^\s*[|`]{2,}", html, re.M), "OK")
log("見出しがh1/h2/h3になっている", html.count("<h1") >= 1 and html.count("<h2") >= 5, f"h1={html.count('<h1')} h2={html.count('<h2')}")
bad_link = re.findall(r'<a[^>]+href="(?!https?:|/|#)[^"]*\.md"', html)
log("READMEへの相対リンクが壊れていない", not bad_link, str(bad_link[:3]))

# ---------- 5) ブラウザでの見た目・コンソールエラー ----------
with sync_playwright() as p:
    b = p.chromium.launch(headless=True)
    pg = b.new_context(viewport={"width": 1280, "height": 900}, locale="ja-JP", device_scale_factor=2).new_page()
    errs = []
    pg.on("console", lambda m: errs.append(m.text[:150]) if m.type == "error" else None)
    pg.goto(DOCS, wait_until="networkidle")
    title = pg.title()
    log("ページタイトルがある", bool(title), title[:50])

    # 目次(ナビ)があるか、リンクが機能するか
    nav = pg.evaluate("() => [...document.querySelectorAll('a[href^=\"#\"]')].length")
    log("見出しへのリンク(目次)がある", nav >= 5, f"{nav}個")

    # 横スクロールが出ていないか(テーブル崩れの検出)
    overflow = pg.evaluate("() => document.documentElement.scrollWidth - document.documentElement.clientWidth")
    log("横スクロールが出ていない", overflow <= 2, f"はみ出し {overflow}px")

    pg.screenshot(path=f"{SHOT}/docs-top.png")
    pg.evaluate("() => window.scrollTo(0, document.body.scrollHeight * 0.35)")
    pg.wait_for_timeout(400)
    pg.screenshot(path=f"{SHOT}/docs-mid.png")
    print(f"   スクショ: {SHOT}/docs-top.png / docs-mid.png")

    log("コンソールエラーなし", not errs, str(errs[:3]))
    b.close()

fails = [x for x in OUT if not x[1]]
print(f"\n=== 合計: {len(OUT) - len(fails)} PASS / {len(fails)} FAIL ===")
for label, _, extra in fails:
    print("  FAIL:", label, extra)
sys.exit(1 if fails else 0)
